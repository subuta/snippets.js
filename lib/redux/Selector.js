import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

import Reducer from './Reducer'

const selectorFor = ({ tableName }, name) => {
  // ensure naming convention
  name = _.camelCase(name)

  const model = _.camelCase(pluralize.singular(tableName))
  const get = _.camelCase('get' + _.upperFirst(name))

  return `const ${get} = state.${model}.${name}`
}

const getAll = ({ tableName }) => {
  // ensure naming convention
  tableName = _.camelCase(pluralize.singular(tableName))

  return build`
  const getAll = createSelector(
    getEntities,
    getIds,
    _.identity,
    (entities, ids, state) => ids.map(id => {
      return denormalize(entities[id], '${tableName}', state)
    })
  )
`
}

// Generate selector for reducers based on named exports.
let selectors = _.transform(_.keys(Reducer), (result, nameOfReducer) => {
  result[nameOfReducer] = (model) => selectorFor(model, nameOfReducer)
}, {})

selectors.getAll = getAll

export default selectors