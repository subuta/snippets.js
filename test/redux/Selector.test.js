/* global expect, describe, it */

import { build, format, snippets as s } from 'bld.js'

import Selector from 'lib/redux/Selector'

describe('redux/Selector', () => {
  it('should create redux entities Selector', () => {
    const code = build`
      export ${Selector.entities({schema: {tableName: 'books'}})}
    `

    expect(format(code)).toMatchSnapshot()
  })

  it('should create redux ids Selector', () => {
    const code = build`
      export ${Selector.ids({schema: {tableName: 'books'}})}
    `

    expect(format(code)).toMatchSnapshot()
  })

  it('should create redux requestProgress Selector', () => {
    const code = build`
      export ${Selector.isRequestProgress({schema: {tableName: 'books'}})}
    `

    expect(format(code)).toMatchSnapshot()
  })

  it('should create redux getAll Selector', () => {
    const code = build`
      export ${Selector.getAll({schema: {tableName: 'books'}})}
    `

    expect(format(code)).toMatchSnapshot()
  })
})
