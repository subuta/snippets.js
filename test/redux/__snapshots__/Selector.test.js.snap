// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/Selector should create redux entities Selector 1`] = `
"export const getEntities = (state) => state.book.entities
"
`;

exports[`redux/Selector should create redux getAll Selector 1`] = `
"export const getAll = createSelector(
  getEntities,
  getIds,
  _.identity,
  (entities, ids, state) =>
    ids.map((id) => {
      return denormalize(entities[id], 'book', state)
    })
)
"
`;

exports[`redux/Selector should create redux ids Selector 1`] = `
"export const getIds = (state) => state.book.ids
"
`;

exports[`redux/Selector should create redux requestProgress Selector 1`] = `
"export const getIsRequestProgress = (state) => state.book.isRequestProgress
"
`;
